### DISCUSSION TITLE
ML & AI base for smart decompilers

### DISCUSSION THREAD

**Discussion by jemo07** [compiler-explorer/compiler-explorer#3875] [Ideas]
Hi, not sure if this has come up, but ever since I saw the site in 2019 I think, It has been in the back of my mind that this is a basis of a great decompiler platform in the making. Consider that code is voluntarily been fed by users, there is a correlation of *code to compiled object, to assembly. This, in most cases is a great opportunity to train and validate an AI decompiler. I would them become a killer paid app, upload a compiled object, then decompile it. with ongoing training it can become almost trivial. I had a project like this, ciphering data from internal project and it worked quite well, you can take it another notch, build IDE with AI validation *static code analysis to warn of potential risks… possibilities are endless. Love the way the site is and would also love to see a cross collier option, ie, given the output as ASSEMBLY take that back to a higher level language.  

Just a thought…

**Comment by partouf** [comment]
That's definitely an interesting idea, but I don't think that's a viable option for us considering the privacy violating implications. In a somewhat similar way in which github Copilot has received its share of complaints about code suggested that was supposed to be private.

**Reply by jemo07** [reply]
Hi, @partouf Sure, I get the privacy concern and I’m no lawyer, so I can’t comment the true implications from a legal basis, but I believe that, just like other environments favor machine-machine metadata, an ML training model ( would do it real time ) would generate a metadata rather than keeping the code, very different that what copilot is doing, slurping an entire repository. I believe that since most data that get uploaded to the site is code-bits ( not the entire program ) might be different enough. I think it might be well worth looking into it, also consider that AI on demand pay is going to be a massive opportunity… SaaS based IDE’s with AI features will be a major App, GPT3 is already doing some of that and you pay on demand by the bitecode, not a lot, but users will be paying 1-5$ a months to regularly go back and review their code. Pen testers, Reverse engineering “Enterprise License” could be larger, but that’s more of a business case issue. Further more, you are not looking at the code itself and leveraging that as a means to create the presentation layer, which is what code pilot does, it has create a library of functions base on other’s code, this feature I’m suggesting would not do such a thing, as the generated presentation will be the outcome of a model that, upon decompiling from opcode to assembly, it will further decompile to a higher level language, it will generate code to describe a function, but not necessarily based on the user code, it can obfuscate function & variable name in the creation of the code. Anyways, I just keep thinking that could be a neat feature… 🤞
